### PR TITLE
Add Y091: Protocol method arg should not be pos-or-kw

### LINKED ISSUE #441: Arguments to protocol methods should never be positional-or-keyword
It's easy to write a protocol like this:

```python
class P(Protocol):
    def write(self, arg: bytes) -> None: ...
```

But it's almost never what you want. Typically you'd use a protocol like this for a function that takes an argument `p` of type `P`. That function probably does `p.write(some_buffer)` or possibly `p.write(arg=some_buffer)`, but almost certainly not both. So the protocol should either have a positional-only or a keyword-only arg, but a positional-or-keyword arg is probably a mistake.

I think it would be reasonable for flake8-pyi to warn about this.

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Fixes #441

### CODE REVIEW

File: ERRORCODES.md
Comment (AlexWaygood) [inline ERRORCODES.md:83]:
```suggestion
```

File: pyi.py
Comment (AlexWaygood) [inline pyi.py:1620]:
```suggestion
```

File: CHANGELOG.md
Comment (JelleZijlstra) [inline CHANGELOG.md:14]:
```suggestion
* Introduce Y062: Protocol method parameters should not be positional-or-keyword.
```

File: ERRORCODES.md
Comment (JelleZijlstra) [reply ERRORCODES.md]: I swear I looked at the diff before making the PR :)

File: pyi.py
Comment (AlexWaygood) [inline pyi.py:2031]: minor micro-optimisation: we don't need to check whether we're in a protocol or not if we know we're not inside a class:

```suggestion
            if self.in_protocol.active:
                self.check_arg_kinds(node)
```

File: pyi.py
Comment (AlexWaygood) [inline pyi.py]: This only allows protocols that use PEP-570 syntax, but typeshed doesn't use any PEP-570 syntax yet.

Also, what if it's a public protocol? In that case, it's probably a stub for a "real" protocol that exists at runtime, so stubtest will start complaining if a parameter is positional-only in the stub but positional-or-keyword at runtime. Maybe we should only complain about private protocols, similar to the way we only complain if a protocol/TypeVar/TypedDict is unused if the protocol/TypeVar/TypedDict is private?

File: ERRORCODES.md
Comment (AlexWaygood) [inline ERRORCODES.md:67]:
```suggestion
| Y062 | Protocol methods should not have positional-or-keyword parameters. Usually, a positional-only parameter is better.
```

File: pyi.py
Comment (JelleZijlstra) [reply pyi.py]: I fixed the code to also handle `__` (typeshed-primer was unhappy with me about that).

Not sure about public protocols; it's probably also a bug in the public code. Let's go over the typeshed hits first to see if there are real-world examples of this problem.

File: tests/protocol_arg.pyi
Comment (AlexWaygood) [inline tests/protocol_arg.pyi:7]: not really necessary in terms of code coverage, but we could maybe also add some examples that use both positional-only and keyword-only args, for completeness:

```suggestion
    def method4(self, __arg: int) -> None: ...
    def method5(self, __arg: int, *, foo: str) -> None: ...
    def method6(self, arg, /, *, foo: str) -> None: ...
```

File: pyi.py
Comment (JelleZijlstra) [inline pyi.py:2002]: This isn't correct for staticmethods. I think it's questionable anyway to have a staticmethod in a protocol, but there's at least one in the standard library (`email.headerregistry._HeaderParser.value_parser`).

Comment (AlexWaygood) [reply pyi.py:2002]: if we want to disallow staticmethods in protocols, I feel like that should probably be a dedicated error code -- I lean towards saying check should probably handle staticmethods in protocols in a principled way?

Comment (JelleZijlstra) [reply pyi.py:2002]: Yes, I should change the code here to handle staticmethods correctly. It's worth thinking about whether we should lint against staticmethods in protocols, but that should be an independent change.

I'm going to hold off on updating this PR for a while though until I can get the typeshed-primer hits down a bit.

File: pyi.py
Comment (hauntsaninja) [inline pyi.py:2257]:
```suggestion
    'Y062 Argument "{arg}" to protocol method "{method}" should probably not be positional-or-keyword. '
    " Make it positional-only, since usually you don't want to mandate a specific argument name"
```

Given that e.g. ruff will reimplement this error, can we soften the error message here

File: pyi.py
Comment (AlexWaygood) [inline pyi.py:1601]: a base cannot be both a `TypedDict` _and_ a `Protocol`:

```suggestion
            elif _is_TypedDict(base):
```

Comment (JelleZijlstra) [reply pyi.py:1601]: I actually don't like that change. The current code is obviously correct; to verify that the elif version is correct, you need to have the background knowledge that a class can't be both a TypedDict and a Protocol.

Comment (AlexWaygood) [reply pyi.py:1601]: Fair enough!

File: pyi.py
Comment (AlexWaygood) [inline pyi.py:2257]: You have two spaces after the full stop currently

```suggestion
    "Make it positional-only, since usually you don't want to mandate a specific argument name"
```

**Comment by AlexWaygood** [comment]
(#449 bagged Y062, so this will have to be Y063 now :)

File: CHANGELOG.md
Comment (AlexWaygood) [inline CHANGELOG.md:86]:
```suggestion
* Introduce Y067: Protocol method parameters should not be positional-or-keyword.
```

File: ERRORCODES.md
Comment (AlexWaygood) [inline ERRORCODES.md:83]:
```suggestion
| <a id="Y067" href="#Y067">Y067</a> | Protocol methods should not have positional-or-keyword parameters. Usually, a positional-only parameter is better.
```

**Comment by JelleZijlstra** [comment]
It's probably best to merge this as an off-by-default rule, and maybe we can eventually build something that alerts on a typeshed PR if new violations are introduced.

File: pyi.py
Comment (JelleZijlstra) [reply pyi.py:2129]: Noting this is still unresolved, should fix before merging

File: pyi.py
Comment (JelleZijlstra) [reply pyi.py]: We now handle both the old and new pos-only syntax. Since this discussion started, typeshed switched to PEP 570 and flake8-pyi started warning about the double underscore syntax, but for this check I think it's still useful to support both, in case some users are still using the old syntax in stubs.

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**
